### PR TITLE
add basic regex auto-reply support

### PR DESCRIPTION
**Pull request by meganrogge**
fixes https://github.com/microsoft/vscode-copilot/issues/16430

Have a new issue  #257726 to track:
- multiple prompts (include this in output polling instead of after)
- support for foreground terminals
- delayed prompts
- ask the llm to determine options vs using regex
- consider automatically replying

### CODE REVIEW

File: src/vs/workbench/contrib/terminalContrib/chatAgentTools/browser/bufferOutputPolling.ts [hunk 1]
```diff
@@ -215,3 +240,37 @@ export async function assessOutputForErrors(buffer: string, token: CancellationT
...
+		{ regex: /\bcontinue\? \(y\/n\)/ig, kind: 'y/n' },
+		{ regex: /\bproceed\? \(y\/n\)/ig, kind: 'y/n' },
+		{ regex: /\btype yes or no\b/ig, kind: 'yes/no' },
+		{ regex: /\b\(yes\/no\)/ig, kind: 'yes/no' },
+		{ regex: /\b\[y\/n\]/ig, kind: 'y/n' },
+		{ regex: /\benter your choice\b/ig, kind: 'choice' },
+		{ regex: /\bselect an option\b/ig, kind: 'choice' },
+		{ regex: /\bplease respond\b/ig, kind: 'response' },
+		{ regex: /\bpress (enter|return|any key)\b/ig, kind: 'press key' }
+	];
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminalContrib/chatAgentTools/browser/bufferOutputPolling.ts]: Did you have programs that output these? We should include the programs in comments if so, so we know the source.

File: src/vs/workbench/contrib/terminalContrib/chatAgentTools/browser/bufferOutputPolling.ts [hunk 2]
```diff
@@ -215,3 +240,37 @@ export async function assessOutputForErrors(buffer: string, token: CancellationT
...
+ * Returns the kind of input expected by the last line of the output if it appears to be a user prompt.
+ * For example, returns 'y/n', 'yes/no', 'press key', 'choice', etc., or undefined if not a prompt.
+ */
+export function getExpectedUserInputKind(output: string): string | undefined {
+	if (!output) {
+		return undefined;
+	}
+
+	const patterns: { regex: RegExp; kind: string }[] = [
+		{ regex: /\bdo you want to continue\b.*\(y\/n\)/ig, kind: 'y/n' },
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminalContrib/chatAgentTools/browser/bufferOutputPolling.ts]: If we idle briefly with one of these matches, we could do more general ones like '/\(y\/n\):?\s*$/', checking the end of the line is important too.

File: src/vs/workbench/contrib/terminalContrib/chatAgentTools/browser/bufferOutputPolling.ts [hunk 3]
```diff
@@ -215,3 +240,37 @@ export async function assessOutputForErrors(buffer: string, token: CancellationT
...
+/**
+ * Returns the kind of input expected by the last line of the output if it appears to be a user prompt.
+ * For example, returns 'y/n', 'yes/no', 'press key', 'choice', etc., or undefined if not a prompt.
+ */
+export function getExpectedUserInputKind(output: string): string | undefined {
+	if (!output) {
+		return undefined;
+	}
+
+	const patterns: { regex: RegExp; kind: string }[] = [
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminalContrib/chatAgentTools/browser/bufferOutputPolling.ts]: This is pwsh's confirm impact one, not sure if there are slight variants but this was from trying to `remove-item folder`.

```
[Y] Yes  [A] Yes to All  [N] No  [L] No to All  [S] Suspend  [?] Help (default is "Y"): 
```

File: src/vs/workbench/contrib/terminalContrib/chatAgentTools/browser/bufferOutputPolling.ts [hunk 4]
```diff
@@ -184,6 +190,25 @@ export function promptForMorePolling(command: string, context: IToolInvocationCo
...
+}
+
+export function promptForYesNo(context: IToolInvocationContext, chatService: IChatService): { promise: Promise<boolean>; part?: ChatElicitationRequestPart } {
+	return createYesNoPrompt(
+		context,
+		chatService,
+		localize('poll.terminal.yes', 'Respond yes in the terminal?'),
+		localize('poll.terminal.yesNo', 'Copilot will run the reply in the terminal.')
+	);
+}
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminalContrib/chatAgentTools/browser/bufferOutputPolling.ts]: Should these live in ./elicitation.ts?

File: src/vs/workbench/contrib/terminalContrib/chatAgentTools/browser/bufferOutputPolling.ts [hunk 5]
```diff
@@ -184,6 +190,25 @@ export function promptForMorePolling(command: string, context: IToolInvocationCo
...
+		title,
+		message
+	);
+}
+
+export function promptForYesNo(context: IToolInvocationContext, chatService: IChatService): { promise: Promise<boolean>; part?: ChatElicitationRequestPart } {
+	return createYesNoPrompt(
+		context,
+		chatService,
+		localize('poll.terminal.yes', 'Respond yes in the terminal?'),
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminalContrib/chatAgentTools/browser/bufferOutputPolling.ts]: We could be more specific:

- Respond "yes" in the terminal?
- Respond "y" in the terminal?

File: src/vs/workbench/contrib/terminalContrib/chatAgentTools/browser/runInTerminalTool.ts [hunk 6]
```diff
@@ -297,14 +297,35 @@ export class RunInTerminalTool extends Disposable implements IToolImpl {
...
 						outputAndIdle,
 						token,
 						this._languageModelsService,
 						execution
 					);
 				}
-
+				const userInputKind = await getExpectedUserInputKind(outputAndIdle.output);
+				if (userInputKind) {
+					if (userInputKind !== 'choice' && userInputKind !== 'key') {
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminalContrib/chatAgentTools/browser/runInTerminalTool.ts]: Can this be reused between tools?

File: src/vs/workbench/contrib/terminalContrib/chatAgentTools/browser/runInTerminalTool.ts [hunk 7]
```diff
@@ -297,14 +297,35 @@ export class RunInTerminalTool extends Disposable implements IToolImpl {
...
-						() => promptForMorePolling(command, invocation.context!, this._chatService),
+						() => promptForMorePolling(localize('poll.terminal.waiting', "Continue waiting for `{0}` to finish?", command), localize('poll.terminal.polling', "Copilot will continue to poll for output to determine when the terminal becomes idle for up to 2 minutes."), invocation.context!, this._chatService),
 						outputAndIdle,
 						token,
 						this._languageModelsService,
 						execution
 					);
 				}
-
+				const userInputKind = await getExpectedUserInputKind(outputAndIdle.output);
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminalContrib/chatAgentTools/browser/runInTerminalTool.ts]: This could happen multiple times and also maybe only after some period of time has passed, I think it needs to live inside the polling loop which should be doing general evaluation of the state instead of just detecting idle.

Comment (Tyriar) [inline src/vs/workbench/contrib/terminalContrib/chatAgentTools/browser/runInTerminalTool.ts]: We want this for foreground terminals too, currently I think they'll just hang and keep spinning until the user responds.

**Review by Tyriar — CHANGES_REQUESTED** [review]

File: src/vs/workbench/contrib/terminalContrib/chatAgentTools/browser/runInTerminalTool.ts (hunk 6 above)
Comment (meganrogge) [reply src/vs/workbench/contrib/terminalContrib/chatAgentTools/browser/runInTerminalTool.ts]: yes, pushed that

File: src/vs/workbench/contrib/terminalContrib/chatAgentTools/browser/task/createAndRunTaskTool.ts [hunk 8]
```diff
@@ -115,13 +115,30 @@ export class CreateAndRunTaskTool implements IToolImpl {
...
+				async () => await pollForOutputAndIdle({ getOutput: () => getOutput(terminal), isActive: () => this._isTaskActive(task) }, true, token, this._languageModelsService),
+			);
+			if (handleResult.handled) {
+				if (handleResult.outputAndIdle) {
+					outputAndIdle = handleResult.outputAndIdle;
+				}
+			} else {
+				return { content: [{ kind: 'text', value: `The task is still running and requires user input of ${userInputKind}.` }], toolResultMessage: new MarkdownString(localize('copilotChat.taskRequiresUserInput', 'The task `{0}` is still running and requires user input. {1}', task._label, userInputKind)) };
+			}
+		}
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminalContrib/chatAgentTools/browser/task/createAndRunTaskTool.ts]: Sharing this between the 3 tools would be good. This could live in elicitation.ts with the prompt calls too?

**Review by Tyriar — DISMISSED** [review]

File: src/vs/workbench/contrib/terminalContrib/chatAgentTools/browser/runInTerminalTool.ts (hunk 7 above)
Comment (meganrogge) [reply src/vs/workbench/contrib/terminalContrib/chatAgentTools/browser/runInTerminalTool.ts]: Deferring

Comment (meganrogge) [reply src/vs/workbench/contrib/terminalContrib/chatAgentTools/browser/runInTerminalTool.ts]: Deferring

File: src/vs/workbench/contrib/terminalContrib/chatAgentTools/browser/task/createAndRunTaskTool.ts (hunk 8 above)
Comment (meganrogge) [reply src/vs/workbench/contrib/terminalContrib/chatAgentTools/browser/task/createAndRunTaskTool.ts]: can't put that in elicitation.ts as there are very specific terminal and task things in this function like outputAndIdle, but moved to bufferOutputPolling

Comment (meganrogge) [reply src/vs/workbench/contrib/terminalContrib/chatAgentTools/browser/task/createAndRunTaskTool.ts]: done though in terms of consolidating them

File: src/vs/workbench/contrib/terminalContrib/chatAgentTools/browser/bufferOutputPolling.ts [hunk 9]
```diff
@@ -152,38 +152,43 @@ export async function pollForOutputAndIdle(
...
+export async function handleTerminalUserInputPrompt(
+	outputAndIdle: { output: string; terminalExecutionIdleBeforeTimeout: boolean; pollDurationMs?: number; modelOutputEvalResponse?: string },
+	invocation: any,
+	chatService: IChatService,
+	terminal: ITerminalInstance,
+	executableLabel: string,
+	token: CancellationToken,
+	languageModelsService: ILanguageModelsService
+): Promise<{ handled: boolean; userResponse?: 'accepted' | 'rejected'; outputAndIdle: { output: string; terminalExecutionIdleBeforeTimeout: boolean; pollDurationMs?: number; modelOutputEvalResponse?: string }; message?: IToolResult }> {
+	const userInputKind = await getExpectedUserInputKind(outputAndIdle.output);
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminalContrib/chatAgentTools/browser/bufferOutputPolling.ts:164]:
```suggestion
	const userInputKind = getExpectedUserInputKind(outputAndIdle.output);
```

**Review by Tyriar — CHANGES_REQUESTED** [review]
Pushed https://github.com/microsoft/vscode/pull/257691/commits/98e45b3c6539870bbf2042fcd60a2b0c7df44454 to add the events we want to collect for this. From my testing the `userResponse` variable doesn't actually work currently.

When I pulled the code I also found it pretty hard to follow. I think it's a little late to push something so large that has the potential to hang the tool due to async logic, especially without any tests.

Since it's an edge case (bg terminals), doesn't handle multiple questions and just does basic regex, I recommend waiting and implementing it in a more robust way next month. It's not worth the risk considering how little this feature actually covers in its current state. What we need is:

- Some output monitoring class that `RunInTerminalTool` hands ownership of the execution to and then it has a loop that does all the idle/question polling and handling in a central place (not inside `RunInTerminalTool.invoke`)
- This component must have a good set of unit tests
- It should return a well structured return object as `handleTerminalUserInputPrompt`'s return result is quite messy right now.

If anything the hint with an a11y alert as I suggested earlier is the safer approach as it's just adding a low risk hint to tell the user the terminal is awaiting input.

**Comment by meganrogge** [comment]
too many conflicts, starting from scratch
